### PR TITLE
Store: Add SQL tables for SQL based object store (dev mode only)

### PR DESCRIPTION
**Pull request by ryantxu**
This PR adds a method to wipe and initialize object store SQL tables while the system is under development.

With this process we can recreate the tables by incrementing a version key:
image

Note the different suffix here:
image

This process will let us ensure that the tables are created optimally before committing them to our long term migration log.

A full featured SQL implementation lives in https://github.com/grafana/grafana/pull/56348, after https://github.com/grafana/grafana/pull/57714 is merged -- we can look to add a more minimal SQL version

### CODE REVIEW

File: pkg/services/sqlstore/migrations/object_store_mig.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+			{Name: "body", Type: migrator.DB_Blob, Nullable: false},
+			{Name: "size", Type: migrator.DB_BigInt, Nullable: false},
+			{Name: "etag", Type: migrator.DB_NVarchar, Length: 32, Nullable: false}, // md5(body)
+			{Name: "version", Type: migrator.DB_NVarchar, Length: 128, Nullable: false},
+
+			// Who changed what when -- We should avoid JOINs with other tables in the database
+			{Name: "updated", Type: migrator.DB_DateTime, Nullable: false},
+			{Name: "created", Type: migrator.DB_DateTime, Nullable: false},
+			{Name: "updated_by", Type: migrator.DB_Int, Nullable: false},
+			{Name: "created_by", Type: migrator.DB_Int, Nullable: false},
```
[the remark this answers is not in the document]
Comment (torkelo) [reply pkg/services/sqlstore/migrations/object_store_mig.go:36]: could be a good idea, easy to store a number in a string, but not the other way around :)

[the remark this answers is not in the document]
Comment (bergquist) [reply pkg/services/sqlstore/migrations/object_store_mig.go:36]: > The gRPC interface should send a string field

I think storage and the grpc should map 1:1 as much as we can.

> easy to store a number in a string, but not the other way around :)

If we use an string  we should be explicit about the fact that it's an identifier and not a display name.

File: pkg/services/sqlstore/migrations/object_store_mig.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		Columns: []*migrator.Column{
+			// Source:
+			{Name: "key", Type: migrator.DB_NVarchar, Length: 1024, Nullable: false},
+
+			// Address (defined in the body, not resolved, may be invalid and change)
+			{Name: "kind", Type: migrator.DB_NVarchar, Length: 255, Nullable: false},
+			{Name: "type", Type: migrator.DB_NVarchar, Length: 255, Nullable: true},
+			{Name: "uid", Type: migrator.DB_NVarchar, Length: 1024, Nullable: true},
+
+			// Runtime calcs (will depend on the system state)
```
Comment (bergquist) [inline pkg/services/sqlstore/migrations/object_store_mig.go:79]: I'm not sure what this is :)

File: pkg/services/sqlstore/migrations/object_store_mig.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	// will only be called when the feature toggle is enabled
+	// this check should not be necessary, but is added as an extra check
+	if setting.Env == setting.Prod {
+		return
+	}
+
+	// Migration cleanups: given that this is a complex setup
+	// that requries a lot of testing before we are ready to push out of dev
+	// this script lets us easy wipe previous changes and initalize clean tables
+	suffix := " (v0)" // change this when we want to wipe and reset the object tables
```
Comment (bergquist) [inline pkg/services/sqlstore/migrations/object_store_mig.go:126]: :+1:

**Review by bergquist — APPROVED** [review]
:+1: lets get something merged. Let's make a long term commitment to the schema right before we have the first real use-case

File: pkg/services/sqlstore/migrations/object_store_mig.go (hunk 2 above)
Comment (ryantxu) [reply pkg/services/sqlstore/migrations/object_store_mig.go:79]: This takes the internal references (eg: type:influx uid:ABC) and says if it resolves ok and to what

**Comment by bergquist** [comment]
:tada:

**Pull request merged**